qutebrowser/browser/mhtml.py

```diff
@@ -0,0 +1,534 @@
+    rels = [rel.lower() for rel in element['rel'].split(' ')]
+    return any(rel in rels for rel in must_have)
+
+
+MHTMLPolicy = email.policy.default.clone(linesep='\r\n', max_line_length=0)
+
+
+# Encode the file using base64 encoding.
+E_BASE64 = email.encoders.encode_base64
+
+
+# Encode the file using MIME quoted-printable encoding.
+E_QUOPRI = email.encoders.encode_quopri
+
+
+class MHTMLWriter():
+
+    """A class for outputting multiple files to a MHTML document.
+
+    Attributes:
+        root_content: The root content as bytes.
+        content_location: The url of the page as str.
+        content_type: The MIME-type of the root content as str.
+        _files: Mapping of location->_File namedtuple.
+    """
+
+    def __init__(self, root_content, content_location, content_type):
+        self.root_content = root_content
+        self.content_location = content_location
+        self.content_type = content_type
+        self._files = {}
+
+    def add_file(self, location, content, content_type=None,
+                 transfer_encoding=E_QUOPRI):
+        """Add a file to the given MHTML collection.
+
+        Args:
+            location: The original location (URL) of the file.
+            content: The binary content of the file.
+            content_type: The MIME-type of the content (if available)
+            transfer_encoding: The transfer encoding to use for this file.
+        """
+        self._files[location] = _File(
+            content=content, content_type=content_type,
+            content_location=location, transfer_encoding=transfer_encoding,
+        )
+
+    def write_to(self, fp):
+        """Output the MHTML file to the given file-like object.
+
+        Args:
+            fp: The file-object, opened in "wb" mode.
+        """
+        msg = email.mime.multipart.MIMEMultipart(
+            'related', '---=_qute-{}'.format(uuid.uuid4()))
+
+        root = self._create_root_file()
+        msg.attach(root)
+
+        for _, file_data in sorted(self._files.items()):
+            msg.attach(self._create_file(file_data))
+
+        gen = email.generator.BytesGenerator(fp, policy=MHTMLPolicy)
+        gen.flatten(msg)
+
+    def _create_root_file(self):
+        """Return the root document as MIMEMultipart."""
+        root_file = _File(
+            content=self.root_content, content_type=self.content_type,
+            content_location=self.content_location, transfer_encoding=E_QUOPRI,
+        )
+        return self._create_file(root_file)
+
+    def _create_file(self, f):
+        """Return the single given file as MIMEMultipart."""
+        msg = email.mime.multipart.MIMEMultipart()
+        msg['Content-Location'] = f.content_location
+        # Get rid of the default type multipart/mixed
+        del msg['Content-Type']
+        if f.content_type:
+            msg.set_type(f.content_type)
+        msg.set_payload(f.content)
+        f.transfer_encoding(msg)
+        return msg
+
+
+class _Downloader():
```

Comment:
I'd prefer if methods here which aren't used from the outside (and which you don't expect to) would be private, i.e. start with `_` - e.g. `_collect_zombies` or `_fetch_url`. I haven't checked if there's more.